File: problemSet/595D-Max_and_Bike.py
"""
Codeforces Round #330 (Div. 2)

Problem 594 B / 595 D. Max and Bike

@author yamaton
@date 2015-11-08
"""

import math
import random
import sys

import functools


@functools.lru_cache(None)
def solve_for_t(r, v, l):
    pm = 1 if (l / 2) % (2 * math.pi * r) < math.pi * r else -1
    def f(t):
        # return v * t + 2 * pm * r * math.sin(half_omega * t) - l
        return v * t + 2 * pm * r * math.sin(v * t / (2 * r)) - l

    def dfdt(t):
        # return v * (1 + pm * math.cos(half_omega * t))
        return v * (1 + pm * math.cos(v * t / (2 * r)))

    def newton(t):
        return t - f(t) / dfdt(t)


    # p('\n----------------')
    estimate = l / v
    prev = estimate
    t = newton(prev)
    while abs(t - prev) > 0.0000001 or abs(t - prev) / max(1, abs(t)) > 0.0000001:
        if t < - estimate or t > 10 * estimate:
            t = random.random() * 2 * estimate
        # p('... t =', t)
        prev, t = t, newton(t)

    # p('prev =', prev)
    # p('left(t) =', v * t + 2 * r * math.sin(v * t / (2 * r)))
    # p('l =', l)
    # p('f(t) =', f(t))

    return t


def solve(pairs, r, v):
    distances = [(f - s) for (s, f) in pairs]
    return [solve_for_t(r, v, d) for d in distances]


def p(*args, **kwargs):
    return print(file=sys.stderr, *args, **kwargs)


def main():
    [n, r, v] = [int(i) for i in input().strip().split()]
    pairs = [[int(i) for i in input().strip().split()] for _ in range(n)]
    results = solve(pairs, r, v)
    for res in results:
        print("%.12f" % res)


if __name__ == '__main__':
    main()

File: problemSet/595D-Max_and_Bike.hs
{-
Codeforces Round #330 (Div. 2)

Problem 594 B / 595 D. Max and Bike

@author yamaton
@date 2015-11-10
-}

import System.Random
import Data.Fixed
import Control.Applicative
import Control.Monad
import Text.Printf
import System.IO (hPutStrLn, stderr)

-- import qualified Data.Map as Map
-- import qualified Data.Set as Set
-- import qualified Data.Foldable as F
-- import qualified Data.Traversable as T

randomSeq :: Int -> [Double]
randomSeq seed = randomRs (0, 1) (mkStdGen seed)

isDiffSmall :: (Double, Double) -> Bool
isDiffSmall (a, b) = diff < 1e-7 && diff / max (abs b)  1.0 < 1e-7
  where diff = abs (a - b)

fixOverShoot :: Double -> Double -> Double -> Double
fixOverShoot lb ub r x
  | lb < x < ub = x
  | otherwise   = lb + (ub - lb) * r

solve :: Int -> Int -> Int -> Double
solve r v d = snd . head $ filter isDiffSmall neighbors
  where
    r' = fromIntegral r
    v' = fromIntegral v
    d' = fromIntegral d
    estimate = d' / v'
    pm = if (d' / 2) `mod'` (2 * pi * r') < pi * r' then 1 else -1

    newton :: Double -> Double
    newton t = t - f t / dfdt t
    f :: Double -> Double
    f t = v' * t + 2 * pm * r' * sin (v' * t / (2 * r')) - d'
    dfdt :: Double -> Double
    dfdt t = v' * (1 + pm * cos (v' * t / 2 * r'))
    lb = -estimate
    ub = 10 * estimate
    rs = randoms 42
    series = scanl (\n r -> (newton . fixOverShoot lb ub r) n) estimate rs
    neighbors = zip series (tail series)


main :: IO ()
main = do
    [n, r, v] <- (map read . words) <$> getLine :: IO [Int]
    forM_ [1..n] $ \_ -> do
      [s, f] <- (map read . words) <$> getLine :: IO [Int]
      print (s, f)
      let d = f - s :: Int
      let result = solve r v d
      print result
      -- hPutStrLn stderr $ printf "d = %s" (show d)
      hPutStrLn stderr (printf "result = %.12f" result)
